rfcs/021-data_science_in_the_pipeline/README.md

```diff
@@ -0,0 +1,54 @@
+
+To resolve this, we want to bring data science services into the pipeline, and use them to augment works/images with data that can be indexed and searched in ES. This RFC details how that might look.
+
+## Proposal
+
+We can write down some desiderata for any proposed solution:
+
+- Data science (DS) logic can be written in Python
+- DS services do not know about our data model(s)
+- Pipeline services do not know about data science
+- DS services do not know about SQS, message passing, the Actor model, etc
+- DS models are persistent and separate to DS inferrers
+- DS models can be retrained on demand
+
+These suggest 3 distinct, but loosely coupled, services:
+
+- **Inferrer**: A Python service that provides a synchronous API (most likely RESTful HTTP) that consumes whatever is needed to infer the new data, which it outputs.
+- **Model Trainer**: A Python service that can consume records from the catalogue index in bulk in order to train a model, and outputs/stores a persistent representation of this model for the *inferrer* to use.
+- **Inference Manager**: A Scala service that lives in the existing pipeline and contains "the usual" Wellcome message-passing, Akka, etc logic & libraries, which sychronously calls the `inferrer` and attaches the new data to the work/image (by populating a field) before passing it along.
+
+The usage of these services would look like this:
+
+![An architecture diagram for data science services](https://user-images.githubusercontent.com/4429247/76087593-631cc700-5fae-11ea-819f-3545e6dc7c65.png)
+
+#### Implementation details
+
+- The inferrer and inference manager exist in one task definition (and therefore on one host). The inferrer is not marked as essential: crashes are tolerated. Timeouts and other failures are naturally handled by the queue redrive policies.
+- The model trainer runs automatically as soon as it starts: it is triggered by an autoscaling policy. Scaling up could be based on regular time intervals, messages in queues, lambdas etc, and scaling down would be done by the trainer itself.
+- The inferrer loads the model from S3 when it starts. Inferrer instances will be short-lived (as they'll scale to zero when not in use) so triggering a restart or loading a new model by other means is not necessary.
+
+### Questions & drawbacks
+
+> Are the network requests for (eg) image files going to be very expensive and/or slow? How could we mitigate this?
+
+> How do we deal with wanting to infer several separate things for one piece of content (eg a feature vector _and_ a palette vector for a single image)? A chain of inferrer-manager pairs would be simplest and cleanest, but this would require `n` times as many network requests for images as one inferrer doing all the work.
```

Comment:
I think this is resolved by treating augmented data as another source.

Reply:
Would it? We'd still have to make the same sort of decisions in that source "generator", I think?

Reply:
> Can we parallel mapreduce the process?

I did wonder about something like this - is it insanity to download images in one place (map) and to send the inferrers big old byte streams (or form multipart uploads) for them to work on (reduce)?

Reply:
My impression is there are a cascade of options for holding image data closer to the inferers depending on how expensive in time & money the process is.

To address parallelising generating feature vectors we could have multiple inferrer services, one for each feature type we need to compute, an inferrer-manager sends messages to multiple inferrers that do their work and report back to the manager which handles completing on success/failure over multiple requests. I think I prefer this to chaining as we can parallelise compute - though perhaps this is not necessary to start with as we'll only have one feature vector to compute?

The inferrers (either as you describe above or as I suggest) could be given access to the image data in a few ways, (in rough order of speed of access):

- Downloading the images from S3 on demand
- Utilising the IIIF API to download smaller images
- Utilising some network accessible cache of those smaller images (possibly utilise the DLCS cache or have our own).
- Utilising some in-memory cache of those smaller image.

The inferrers could be given more access to host hardware by moving away from Fargate and doing something a bit more exotic with an EC2 cluster host too.

I'd start with the easy solution (getting little images from DLCS and utilising its cache), and move on to more interesting infrastructure as required.

Reply:
Oop, and I forgot batch processing stuff here is an option too.